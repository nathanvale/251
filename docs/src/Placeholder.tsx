import {docs} from '../../packages/ods-core/src/Placeholder/Placeholder.docs';
import {renderDocs} from './_private/renderDocs';

export const PlaceholderDocument = () => renderDocs(docs);
