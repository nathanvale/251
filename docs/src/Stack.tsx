import React from 'react';
import {Stack, StackProps, Placeholder} from '@origin-digital/ods-core';
import {Props, Example, Page} from './_private';
import {ContentSection} from '@origin-digital/ods-lab';

export const StackDocument = () => (
  <Page>
    <Props<StackProps>
      title="<Stack/>"
      componentName="Stack"
      description="Stack is a low-level atomic layout component. Its main concerns are vertically stacking and evenly distributing white space between its children. Open the examples in Playroom to see the responsive behaviour of this component."
      propDescriptions={{}}
    />
    <Stack>
      <Example
        title="Default Behaviour"
        description="By default <Stack/> vertically stacks siblings without any space between them."
        Code={() => (
          <Stack>
            <Placeholder backgroundColor="red" />
            <Placeholder backgroundColor="orange" />
            <Placeholder backgroundColor="blue" />
          </Stack>
        )}
      />
      <Example
        title="Distributed Space"
        description={`In order to distribute white space evenly between components pass in "space" with a tshirt size.`}
        Code={() => (
          <Stack space="xxlarge">
            <Placeholder backgroundColor="red" />
            <Placeholder backgroundColor="orange" />
            <Placeholder backgroundColor="blue" />
          </Stack>
        )}
      />
      <Example
        title="Responsive Space on Mobile and Desktop Breakpoints"
        description="Distributed space can be controlled responsively by passing in an array of exactly two tshirt sizes - one for mobile and one for desktop. Open in Playroom to see the responsive behaviour of this example."
        Code={() => (
          <Stack space={['none', 'large']}>
            <Placeholder backgroundColor="red" />
            <Placeholder backgroundColor="orange" />
            <Placeholder backgroundColor="blue" />
          </Stack>
        )}
      />
      <Example
        title="Responsive Space Across All Five Breakpoints"
        description="For more granular control across all 5 breakpoints, xs, sm, md, lg and xl, distributed space can be controlled responsively by passing in an object with breakpoints as keys and tshirt sizes as values. Open in Playroom to see the responsive behaviour of this example."
        Code={() => (
          <Stack
            space={{
              xs: 'none',
              sm: 'small',
              md: 'medium',
              lg: 'large',
              xl: 'xxlarge',
            }}
          >
            <Placeholder backgroundColor="red" />
            <Placeholder backgroundColor="orange" />
            <Placeholder backgroundColor="blue" />
          </Stack>
        )}
      />
      <Example
        title="Dividers"
        description={`Dividers between distributed children can be made visible when you pass in "dividers" as true.`}
        Code={() => (
          <ContentSection backgroundColor="white">
            <Stack space="small" dividers>
              <Placeholder backgroundColor="red" />
              <Placeholder backgroundColor="orange" />
              <Placeholder backgroundColor="blue" />
            </Stack>
          </ContentSection>
        )}
      />
      <Example
        title="Horizontal Alignment"
        description={`With alignX one can decide to align the children of a Stack component, to left, right, center or stretch (default).`}
        Code={() => (
          <ContentSection backgroundColor="white">
            <Stack space="small" dividers alignX="center">
              <Placeholder backgroundColor="red" />
              <Placeholder backgroundColor="orange" />
              <Placeholder backgroundColor="blue" />
            </Stack>
          </ContentSection>
        )}
      />
    </Stack>
  </Page>
);
