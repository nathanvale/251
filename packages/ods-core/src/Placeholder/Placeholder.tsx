import React from "react";
import styled, { css } from "styled-components";
import { width, height, WidthProps, HeightProps } from "styled-system";
import { ResponsiveProp } from "@origin-digital/ods-types";
import { normaliseResponsiveProp } from "@origin-digital/ods-helpers";
import { Box, BoxProps } from "../Box/Box";

type PlaceholderShapeVariant = "rectangle" | "round";
export interface PlaceholderProps {
  "data-id"?: string;
  height?: ResponsiveProp<string | number>;
  label?: string;
  shape?: PlaceholderShapeVariant;
  width?: ResponsiveProp<string | number>;
}

interface ContainerProps extends WidthProps, HeightProps {
  shape: PlaceholderProps["shape"];
}

const Container = styled<Omit<BoxProps, "width" | "height">>(Box)<
  ContainerProps
>`
  ${width}
  ${height}
  border: 2px solid rgba(51, 51, 51, 0.3);
  ${p =>
    p.shape === "round"
      ? css`
          border-radius: 50%;
        `
      : undefined}}
`;

const defaultHeight = 120;
const defaultWidth = "auto";
const defaultShape = "rectangle";

export const Placeholder = ({
  "data-id": dataId,
  label,
  width = defaultWidth,
  height = defaultHeight,
  shape = defaultShape,
}: PlaceholderProps) => {
  return (
    <Container
      data-id={dataId}
      position="relative"
      overflow="hidden"
      display="flex"
      alignItems="center"
      justifyContent="center"
      width={normaliseResponsiveProp<string | number>(width)}
      height={normaliseResponsiveProp<string | number>(height)}
      shape={shape}
      backgroundColor="grey16"
    >
      {label ? (
        <Box
          paddingX="xsmall"
          paddingY="xxsmall"
          textAlign="center"
          style={{ marginTop: "2px" }}
        >
          {label}
        </Box>
      ) : (
        <svg
          xmlns="http://www.w3.org/2000/svg"
          style={{ width: "100%", height: "100%", position: "absolute" }}
        >
          <line
            style={{ strokeWidth: "2px", stroke: "rgba(51,51,51,.1)" }}
            x1={0}
            y1={0}
            x2="100%"
            y2="100%"
          />
          <line
            style={{ strokeWidth: "2px", stroke: "rgba(51,51,51,.1)" }}
            x1="100%"
            y1={0}
            x2={0}
            y2="100%"
          />
        </svg>
      )}
    </Container>
  );
};

Placeholder.defaultProps = {
  "data-id": "placeholder",
  height: defaultHeight,
  inline: false,
  shape: defaultShape,
  width: defaultWidth,
};

Placeholder.displayName = "Placeholder";
