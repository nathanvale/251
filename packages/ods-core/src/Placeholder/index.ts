export { Placeholder } from "./Placeholder";
export type PlaceholderProps = import("./Placeholder").PlaceholderProps;
