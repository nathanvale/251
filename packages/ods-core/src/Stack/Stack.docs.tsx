/* eslint-disable react/display-name */
import React, { ReactNode } from "react";
import { ComponentDocs } from "@origin-digital/ods-types";
import { Box, Section, Stack, StackProps, Placeholder } from "../";

const Container = ({ children }: { children: ReactNode }) => (
  <Box width="full" backgroundColor="white" paddingY={["xlarge", "xxlarge"]}>
    <Section>{children}</Section>
  </Box>
);

export const docs: ComponentDocs<StackProps> = {
  category: "Layout",
  componentName: "Stack",
  description:
    "Stack is a low-level atomic layout component. Its main concerns are vertically stacking and evenly distributing white space between its children. Open the examples in Playroom to see the responsive behaviour of this component.",
  propDescriptions: {},
  migrationGuide: false,
  examples: [
    {
      label: "Default Behaviour",
      description:
        "By default <Stack/> vertically stacks siblings without any space between them.",
      Code: () => (
        <Stack>
          <Placeholder />
          <Placeholder />
          <Placeholder />
        </Stack>
      ),
    },
    {
      label: "Distributed Space",
      description: `In order to distribute white space evenly between components pass in "space" with a tshirt size.`,
      Code: () => (
        <Stack space="xxlarge">
          <Placeholder />
          <Placeholder />
          <Placeholder />
        </Stack>
      ),
    },
    {
      label: "Responsive Space Across All Five Breakpoints",
      description:
        "For more granular control across all 5 breakpoints, xs, sm, md, lg and xl, distributed space can be controlled responsively by passing in an object with breakpoints as keys and tshirt sizes as values. Open in Playroom to see the responsive behaviour of this example.",
      Code: () => (
        <Stack
          space={{
            xs: "none",
            sm: "small",
            md: "medium",
            lg: "large",
            xl: "xxlarge",
          }}
        >
          <Placeholder />
          <Placeholder />
          <Placeholder />
        </Stack>
      ),
    },
    {
      label: "Dividers",
      description: `Dividers between distributed children can be made visible when you pass in "dividers" as true.`,
      Container,
      Code: () => (
        <Stack space="small" dividers>
          <Placeholder />
          <Placeholder />
          <Placeholder />
        </Stack>
      ),
    },
    {
      label: "Horizontal Alignment",
      description:
        "With alignX one can decide to align the children of a Stack component, to left, right, center or stretch (default).",
      Container,
      Code: () => (
        <Stack space="small" dividers alignX="center">
          <Placeholder shape="round" width="120px" />
          <Placeholder shape="round" width="120px" />
          <Placeholder shape="round" width="120px" />
        </Stack>
      ),
    },
  ],
  snippets: [
    {
      label: "XXSmall Space",
      Code: () => (
        <Stack space="xxsmall">
          <Placeholder height={40} />
          <Placeholder height={40} />
          <Placeholder height={40} />
        </Stack>
      ),
    },
    {
      label: "XSmall Space",
      Code: () => (
        <Stack space="xsmall">
          <Placeholder height={40} />
          <Placeholder height={40} />
          <Placeholder height={40} />
        </Stack>
      ),
    },
    {
      label: "Small Space",
      Code: () => (
        <Stack space="small">
          <Placeholder height={40} />
          <Placeholder height={40} />
          <Placeholder height={40} />
        </Stack>
      ),
    },
    {
      label: "Medium Space",
      Code: () => (
        <Stack space="medium">
          <Placeholder height={40} />
          <Placeholder height={40} />
          <Placeholder height={40} />
        </Stack>
      ),
    },
    {
      label: "Large Space",
      Code: () => (
        <Stack space="large">
          <Placeholder height={40} />
          <Placeholder height={40} />
          <Placeholder height={40} />
        </Stack>
      ),
    },
    {
      label: "Responsive Space 1",
      Code: () => (
        <Stack space={["small", "xxxlarge"]}>
          <Placeholder height={40} />
          <Placeholder height={40} />
          <Placeholder height={40} />
        </Stack>
      ),
    },
    {
      label: "Responsive Space 2",
      Code: () => (
        <Stack
          space={{
            xs: "none",
            sm: "small",
            md: "medium",
            lg: "large",
            xl: "xxxlarge",
          }}
        >
          <Placeholder height={40} />
          <Placeholder height={40} />
          <Placeholder height={40} />
        </Stack>
      ),
    },
  ],
};
