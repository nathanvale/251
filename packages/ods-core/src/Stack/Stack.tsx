import React, {ReactNode, Children} from "react";
import styled from "styled-components";
import {AlignItemsVariants} from "@origin-digital/ods-types";
import {Box, ResponsiveSpace} from "../Box/Box";
import {Divider} from "../Divider/Divider";

export type AlignXType = "left" | "center" | "right" | "stretch";

const mapHAlignToAlignItems = (alignX: AlignXType): AlignItemsVariants => {
  const map = {
    left: "flex-start",
    center: "center",
    right: "flex-end",
    stretch: "stretch",
  } as {[k in AlignXType]: AlignItemsVariants};
  return map[alignX];
};

export interface StackProps {
  children: ReactNode;
  space?: ResponsiveSpace;
  alignX?: AlignXType;
  dividers?: boolean;
  "data-id"?: string;
}

const StyledBox = styled(Box)`
  &:last-child {
    padding-bottom: 0;
  }
`;

export const Stack = ({
  children,
  space,
  dividers,
  "data-id": dataId,
  alignX = "stretch",
}: StackProps) => {
  const stackItems = Children.toArray(children);

  return (
    <Box
      data-id={dataId}
      width="full"
      display="flex"
      flexDirection="column"
      alignItems={mapHAlignToAlignItems(alignX)}
    >
      {stackItems.map((child, index) =>
        dividers ? (
          <React.Fragment key={index}>
            {index > 0 ? (
              <Box alignSelf="stretch" paddingY={space}>
                <Divider />
              </Box>
            ) : null}
            <Box>{child}</Box>
          </React.Fragment>
        ) : (
          <StyledBox paddingBottom={space} key={index}>
            {child}
          </StyledBox>
        ),
      )}
    </Box>
  );
};

Stack.defaultProps = {
  "data-id": "stack",
  dividers: false,
  alignX: "stretch",
  space: "none",
};

Stack.displayName = "Stack";
