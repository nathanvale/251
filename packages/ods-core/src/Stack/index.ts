export { Stack, StackChild } from "./Stack";
export type StackProps = import("./Stack").StackProps;
export type StackChildProps = import("./Stack").StackChildProps;
