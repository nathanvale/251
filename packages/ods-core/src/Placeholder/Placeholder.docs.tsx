import React from "react";
import { ComponentDocs } from "@origin-digital/ods-types";
import { Placeholder, PlaceholderProps } from "./Placeholder";

export const docs: ComponentDocs<PlaceholderProps> = {
  category: "Layout",
  componentName: "Placeholder",
  description:
    "This component is helpful for laying out placeholders for future content.",
  propDescriptions: {},
  migrationGuide: false,
  examples: {
    default: {
      Code: () => <Placeholder />,
    },
    additional: [
      {
        label: "Setting the Label",
        Code: () => <Placeholder label="Placeholder for content" />,
      },
      {
        label: "Setting the Shape",
        Code: () => <Placeholder shape="round" width={120} height={120} />,
      },
      {
        label: "Setting the Width",
        Code: () => <Placeholder width={120} />,
      },
      {
        label: "Setting the Height",
        Code: () => <Placeholder height={400} />,
      },
      {
        label: "Responsive Height Across Mobile and Desktop",
        description:
          "Width can be controlled responsively by passing in an array of mobile and desktop pixel sizes as values. Open in Playroom to see the responsive behaviour of this example.",
        Code: () => <Placeholder height={[100, 500]} />,
      },
      {
        label: "Responsive Height Across All Five Breakpoints",
        description:
          "For more granular control across all 5 breakpoints, ‘xs’, ‘sm’,  ‘md’, ‘lg’ and ‘xl’, width can be controlled responsively by passing in an object with breakpoints as keys and pixel sizes as values. Open in Playroom to see the responsive behaviour of this example.",
        Code: () => (
          <Placeholder
            height={{
              xs: 100,
              sm: 200,
              md: 300,
              lg: 400,
              xl: 500,
            }}
          />
        ),
      },
    ],
  },
  snippets: [],
};
