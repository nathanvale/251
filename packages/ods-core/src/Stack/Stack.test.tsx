import React from "react";
import {
  render,
  queryAllByAttribute,
} from "@origin-digital/ods-testing-library";
import { Placeholder } from "../Placeholder/Placeholder";
import { Stack } from "./Stack";

test("It can stack", () => {
  const { container } = render(
    <Stack>
      <Placeholder />
      <Placeholder />
      <Placeholder />
    </Stack>,
  );
  expect(container.firstChild).toMatchSnapshot();
});

test("It can distribute space", () => {
  const { container } = render(
    <Stack space="xxlarge">
      <Placeholder />
      <Placeholder />
      <Placeholder />
    </Stack>,
  );

  const stackChildren = container.firstChild!.childNodes;

  // All children except the last one have padding-bottom set to the value of space prop.
  for (let childIdx = 0; childIdx < stackChildren.length - 1; childIdx++) {
    const wrapperEl = stackChildren[childIdx];
    const style = window.getComputedStyle(wrapperEl as Element);
    expect(style.marginBottom).toEqual("48px");
  }

  // last element has padding-bottom set to 0.
  const style = window.getComputedStyle(stackChildren[
    stackChildren.length - 1
  ] as Element);
  expect(style.paddingBottom).toEqual("0px");
});

test("It adds dividers", () => {
  const { container } = render(
    <Stack space="xxlarge" dividers>
      <Placeholder />
      <Placeholder />
      <Placeholder />
    </Stack>,
  );
  expect(container.firstChild).toMatchSnapshot();
});
test("It does not add dividers if there is only 1 child", () => {
  const { container } = render(
    <Stack space="medium" dividers>
      <Placeholder />
    </Stack>,
  );
  expect(container.children).toHaveLength(1);
});

test("It stretches children by default", () => {
  const { container } = render(
    <Stack>
      <Placeholder />
      <Placeholder />
    </Stack>,
  );

  const columnsEl = queryAllByAttribute("data-id", container, "stack")[0];
  const style = window.getComputedStyle(columnsEl as Element);
  expect(style.alignItems).toEqual("stretch");
});

test("It can right-align components", () => {
  const { container } = render(
    <Stack alignX="right">
      <Placeholder width="120px" />
      <Placeholder width="120px" />
    </Stack>,
  );
  const columnsEl = queryAllByAttribute("data-id", container, "stack")[0];
  const style = window.getComputedStyle(columnsEl as Element);
  expect(style.alignItems).toEqual("flex-end");
});

test("It can center-align components", () => {
  const { container } = render(
    <Stack alignX="center">
      <Placeholder width="120px" />
      <Placeholder width="120px" />
    </Stack>,
  );
  const columnsEl = queryAllByAttribute("data-id", container, "stack")[0];
  const style = window.getComputedStyle(columnsEl as Element);
  expect(style.alignItems).toEqual("center");
});

test("It can left-align children", () => {
  const { container } = render(
    <Stack alignX="left">
      <Placeholder data-id="childElement" width="120px" />
    </Stack>,
  );
  const columnsEl = queryAllByAttribute("data-id", container, "stack")[0];
  const style = window.getComputedStyle(columnsEl as Element);
  expect(style.alignItems).toEqual("flex-start");
});

test("It adds data-id", () => {
  const dom = render(
    <Stack data-id="MyCustomStack">
      <Placeholder />
      <Placeholder />
    </Stack>,
  );

  expect(
    queryAllByAttribute("data-id", dom.container, "MyCustomStack"),
  ).toHaveLength(1);
});
