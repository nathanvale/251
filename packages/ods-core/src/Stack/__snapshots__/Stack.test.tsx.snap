// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It adds dividers 1`] = `
.c2 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  background-color: #e3e3e3;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  overflow: hidden;
}

.c3 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  position: relative;
}

.c5 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  position: absolute;
  width: 100%;
}

.c1 {
  width: auto;
  height: 120px;
  border: 2px solid rgba(51,51,51,0.3);
}

.c4 {
  height: 1px;
  box-shadow: inset 0 0 0 1px #e3e3e3;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c0.c0.c0 > *:not(:last-child) {
  margin-bottom: 48px;
}

<div
  class="c0"
  data-id="stack"
>
  <div
    class="c1 c2"
    data-id="placeholder"
    display="flex"
    height="120"
    overflow="hidden"
    shape="rectangle"
    width="auto"
  >
    <svg
      style="width: 100%; height: 100%; position: absolute;"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
        x1="0"
        x2="100%"
        y1="0"
        y2="100%"
      />
      <line
        style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
        x1="100%"
        x2="0"
        y1="0"
        y2="100%"
      />
    </svg>
  </div>
  <div
    class="c3"
    data-id="divider"
  >
    <div
      class="c4 c5"
      width="full"
    />
  </div>
  <div
    class="c1 c2"
    data-id="placeholder"
    display="flex"
    height="120"
    overflow="hidden"
    shape="rectangle"
    width="auto"
  >
    <svg
      style="width: 100%; height: 100%; position: absolute;"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
        x1="0"
        x2="100%"
        y1="0"
        y2="100%"
      />
      <line
        style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
        x1="100%"
        x2="0"
        y1="0"
        y2="100%"
      />
    </svg>
  </div>
  <div
    class="c3"
    data-id="divider"
  >
    <div
      class="c4 c5"
      width="full"
    />
  </div>
  <div
    class="c1 c2"
    data-id="placeholder"
    display="flex"
    height="120"
    overflow="hidden"
    shape="rectangle"
    width="auto"
  >
    <svg
      style="width: 100%; height: 100%; position: absolute;"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
        x1="0"
        x2="100%"
        y1="0"
        y2="100%"
      />
      <line
        style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
        x1="100%"
        x2="0"
        y1="0"
        y2="100%"
      />
    </svg>
  </div>
</div>
`;

exports[`It can stack 1`] = `
.c2 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  background-color: #e3e3e3;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  overflow: hidden;
}

.c1 {
  width: auto;
  height: 120px;
  border: 2px solid rgba(51,51,51,0.3);
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c0.c0.c0 > *:not(:last-child) {
  margin-bottom: 0px;
}

<div
  class="c0"
  data-id="stack"
>
  <div
    class="c1 c2"
    data-id="placeholder"
    display="flex"
    height="120"
    overflow="hidden"
    shape="rectangle"
    width="auto"
  >
    <svg
      style="width: 100%; height: 100%; position: absolute;"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
        x1="0"
        x2="100%"
        y1="0"
        y2="100%"
      />
      <line
        style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
        x1="100%"
        x2="0"
        y1="0"
        y2="100%"
      />
    </svg>
  </div>
  <div
    class="c1 c2"
    data-id="placeholder"
    display="flex"
    height="120"
    overflow="hidden"
    shape="rectangle"
    width="auto"
  >
    <svg
      style="width: 100%; height: 100%; position: absolute;"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
        x1="0"
        x2="100%"
        y1="0"
        y2="100%"
      />
      <line
        style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
        x1="100%"
        x2="0"
        y1="0"
        y2="100%"
      />
    </svg>
  </div>
  <div
    class="c1 c2"
    data-id="placeholder"
    display="flex"
    height="120"
    overflow="hidden"
    shape="rectangle"
    width="auto"
  >
    <svg
      style="width: 100%; height: 100%; position: absolute;"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
        x1="0"
        x2="100%"
        y1="0"
        y2="100%"
      />
      <line
        style="stroke-width: 2px; stroke: rgba(51,51,51,.1);"
        x1="100%"
        x2="0"
        y1="0"
        y2="100%"
      />
    </svg>
  </div>
</div>
`;
